azuredevops/internal/service/git/resource_git_repository_file.go

```diff
@@ -0,0 +1,379 @@
+			"file": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The file path to manage",
+			},
+			"content": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The file's content",
+			},
+			"branch": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The branch name, defaults to \"master\"",
+				Default:     "refs/heads/master",
+			},
+			"commit_message": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The commit message when creating or updating the file",
+			},
+			"overwrite_on_create": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Enable overwriting existing files, defaults to \"false\"",
+				Default:     false,
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(1 * time.Minute),
+			Read:   schema.DefaultTimeout(5 * time.Second),
+		},
+	}
+}
+
+func resourceGitRepositoryPushArgs(d *schema.ResourceData, objectID string, changeType git.VersionControlChangeType) (*git.CreatePushArgs, error) {
+	var message *string
+	if commitMessage, hasCommitMessage := d.GetOk("commit_message"); hasCommitMessage {
+		cm := commitMessage.(string)
+		message = &cm
+	}
+
+	repo := d.Get("repository_id").(string)
+	content := d.Get("content").(string)
+	file := d.Get("file").(string)
+	branch := d.Get("branch").(string)
+
+	change := git.GitChange{
+		ChangeType: &changeType,
+		Item: git.GitItem{
+			Path: &file,
+		},
+		NewContent: &git.ItemContent{
+			Content:     &content,
+			ContentType: &git.ItemContentTypeValues.RawText,
+		},
+	}
+
+	args := &git.CreatePushArgs{
+		RepositoryId: &repo,
+		Push: &git.GitPush{
+			RefUpdates: &[]git.GitRefUpdate{
+				{
+					Name:        &branch,
+					OldObjectId: &objectID,
+				},
+			},
+			Commits: &[]git.GitCommitRef{
+				{
+					Comment: message,
+					Changes: &[]interface{}{change},
+				},
+			},
+		},
+	}
+
+	return args, nil
+}
+
+func resourceGitRepositoryFileCreate(d *schema.ResourceData, m interface{}) error {
+	ctx := context.Background()
+	clients := m.(*client.AggregatedClient)
+
+	repo := d.Get("repository_id").(string)
+	file := d.Get("file").(string)
+	branch := d.Get("branch").(string)
+	overwriteOnCreate := d.Get("overwrite_on_create").(bool)
+
+	if err := checkRepositoryBranchExists(clients, repo, branch); err != nil {
+		return err
+	}
+
+	changeType := git.VersionControlChangeTypeValues.Add
+	item, err := clients.GitReposClient.GetItem(ctx, git.GetItemArgs{
+		RepositoryId: &repo,
+		Path:         &file,
+	})
+	if err != nil && !utils.ResponseWasNotFound(err) {
+		return err
+	}
+
+	if item != nil {
+		if !overwriteOnCreate {
+			return fmt.Errorf("Refusing to overwrite existing file. Configure `overwrite_on_create` to `true` to override.")
+		} else {
+			changeType = git.VersionControlChangeTypeValues.Edit
+		}
+	}
```

Comment:
```suggestion
	if item != nil && !overwriteOnCreate {
		return fmt.Errorf("Refusing to overwrite existing file. Configure `overwrite_on_create` to `true` to override.")
	}
	changeType = git.VersionControlChangeTypeValues.Edit
```

Reply:
@xuzhang3 the logic you have suggested here is different to the original logic:

1. if fails the test TestAccGitRepoFile_CreateAndUpdate
2. changeType = git.VersionControlChangeTypeValues.Edit always gets run when line 161 is not run

Test result:
```
API server listening at: 127.0.0.1:42024
=== RUN   TestAccGitRepoFile_CreateAndUpdate
    testing.go:673: Step 0 error: errors during apply:
        
        Error: The path 'foo.txt' does not exist at commit '64fba23b7b92e632619b0758c1eeb059fa685425'
        Parameter name: newPush
        
          on /tmp/tf-test256550528/main.tf line 2:
          (source code not available)
        
        
--- FAIL: TestAccGitRepoFile_CreateAndUpdate (15.91s)
FAIL
```

Reply:
Yes you are right about this one. Idiomatic go seems to be about never using else statements so I will change this as the else statement is not required after the error is returned.